src/mpid/ch4/src/ch4r_symheap.h

```diff
@@ -193,6 +193,236 @@ static inline void *MPIDI_CH4R_generate_random_addr(size_t size)
+            char serialized_hnd[MPL_SHM_GHND_SZ] = { 0 };
+
+            /* get serialized handle from rank 0 and deserialize it */
+            mpi_errno = MPIR_Bcast(serialized_hnd, MPL_SHM_GHND_SZ, MPI_CHAR, 0,
+                                   shm_comm_ptr, &errflag);
+            MPIR_ERR_CHKANDJUMP(errflag, mpi_errno, MPI_ERR_OTHER, "**coll_fail");
+
+            mpl_err = MPL_shm_hnd_deserialize(*shm_segment_hdl_ptr, serialized_hnd,
+                                              strlen(serialized_hnd));
+            MPIR_ERR_CHKANDJUMP(mpl_err != MPL_SHM_SUCCESS, mpi_errno, MPI_ERR_OTHER,
+                                "**alloc_shar_mem");
+
+            /* try to attach with specified symmetric address */
+            mpl_err = MPL_shm_fixed_seg_attach(*shm_segment_hdl_ptr, shm_segment_len, base_ptr, 0);
+            if (mpl_err == MPL_SHM_EINVAL)
+                *mapfail_flag = 1;
+
+            /* handle mapping failure separately, caller may retry predefined times before giving up */
+            MPIR_ERR_CHKANDJUMP((mpl_err != MPL_SHM_EINVAL && mpl_err != MPL_SHM_SUCCESS),
+                                mpi_errno, MPI_ERR_OTHER, "**alloc_shar_mem");
+
+            mpi_errno = MPIR_Barrier(shm_comm_ptr, &errflag);
+            MPIR_ERR_CHKANDJUMP(errflag, mpi_errno, MPI_ERR_OTHER, "**coll_fail");
+        }
+    }
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_CH4I_ALLOCATE_SYMSHM_SEGMENT);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+/* Allocate a symmetric heap over a communicator with same starting address
+ * on each process and shared memory over processes of each node. The MPL_shm
+ * routines are internally used on node with multiple processes, and MPL_mmap
+ * is used on node with single process. It tries MPIR_CVAR_CH4_SHM_SYMHEAP_RETRY
+ * times before giving up, and returns with fail_flag=1 if mapping fails.
+ *
+ * Input shm_size is the total mapping size of shared memory on local node, and
+ * shm_offsets presents the expected offset of local process's start address in
+ * the mapped shared memory. */
+#undef FUNCNAME
+#define FUNCNAME MPIDI_CH4R_get_shm_symheap
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+static inline int MPIDI_CH4R_get_shm_symheap(MPI_Aint shm_size, MPI_Aint * shm_offsets,
+                                             MPIR_Comm * comm, MPIR_Win * win, int *fail_flag)
+{
+    int mpi_errno = MPI_SUCCESS;
+    int iter = MPIR_CVAR_CH4_SHM_SYMHEAP_RETRY;
+    unsigned any_mapfail_flag = 1;
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_CH4R_GET_SHM_SYMHEAP);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_CH4R_GET_SHM_SYMHEAP);
+
+#ifdef USE_SYM_HEAP
+    MPL_shm_hnd_t *shm_segment_hdl_ptr = &MPIDI_CH4U_WIN(win, shm_segment_handle);
+    void **base_ptr = &MPIDI_CH4U_WIN(win, mmap_addr);
+
+    size_t mapsize = 0, page_sz;
+    MPIR_Errflag_t errflag = MPIR_ERR_NONE;
+    MPIR_Comm *shm_comm_ptr = comm->node_comm;
+
+    struct {
+        size_t sz;
```

Comment:
I just reused the code from old symheap allocation. But you are right that the types seems not match. How about replace it with `MPI_Aint` ?

Reply:
Also, if you are to change this to `long`, you'll need to be careful about the code like this.
https://github.com/pmodels/mpich/pull/3319/files/101f448325a0e15acff2e8c3efc7831d3a3d9bc1..d7faf7a4119d2f18593f152a0cd437fda14638c9#diff-4a7415a94d2f7728e88418ad6a750a76R336

```C
maxloc.sz = mapsize;
```
will no longer be correct because you may be truncating the value.

Another idea would be to define a user-defined operator that supports `MAXLOC` for `size_t`?

Reply:
We might want to change the type of all dependent parameters. `MPI_Aint` represents address value, which should be safer than `long`. It can be also directly translated to `MPI_AINT` in the reduce operation.